File: space-game/Data.SqlDataProvider/Extensions/IQueryableExtensions.vb
Imports System.Runtime.CompilerServices
Imports System.Data.Entity

Namespace Extensions
    Public Module QueryableExtensions

        <Extension>
        Public Function IncludeAllTables(this As IQueryable(Of CelestialObject)) As IQueryable(Of CelestialObject)
            Return this.
                IncludeType().
                IncludeRingSystem().
                IncludeTextures()

        End Function

        <Extension>
        Private Function IncludeTextures(this As IQueryable(Of CelestialObject)) As IQueryable(Of CelestialObject)
            Return this.Include(Function(s) s.TextureGroup).
                                Include(Function(s) s.TextureGroup).
                                Include(Function(s) s.TextureGroup.TextureGroupToTextures.Select(Function(t) t.Texture.TextureType)).
                                Include(Function(s) s.TextureGroup.TextureGroupToTextures.Select(Function(t) t.Texture.TexturePath)).
                                Include(Function(s) s.TextureGroup.TextureGroupToTextures.Select(Function(t) t.Texture.TextureQuality))
        End Function

        <Extension>
        Private Function IncludeType(this As IQueryable(Of CelestialObject)) As IQueryable(Of CelestialObject)
            Return this.Include(Function(s) s.CelestialObjectType)
        End Function

        <Extension>
        Private Function IncludeRingSystem(this As IQueryable(Of CelestialObject)) As IQueryable(Of CelestialObject)
            Return this.Include(Function(s) s.RingSystem).
                                Include(Function(s) s.RingSystem.TextureGroup).
                                Include(Function(s) s.RingSystem.TextureGroup.TextureGroupToTextures.Select(Function(t) t.Texture.TextureType)).
                                Include(Function(s) s.RingSystem.TextureGroup.TextureGroupToTextures.Select(Function(t) t.Texture.TexturePath)).
                                Include(Function(s) s.RingSystem.TextureGroup.TextureGroupToTextures.Select(Function(t) t.Texture.TextureQuality))
        End Function

    End Module
End Namespace

File: space-game/Data.SqlDataProvider/Extensions/IQueryableExtensions.cs
using Data.SqlDataProvider.Model;
using System.Data.Entity;
using System.Linq;

namespace Data.SqlDataProvider.Extensions
{
    // ReSharper disable once InconsistentNaming
    public static class IQueryableExtensions
    {
        public static IQueryable<CelestialObject> IncludeAllTables(this IQueryable<CelestialObject> @this)
        {
            return @this.IncludeType()
                        .IncludeRingSystem()
                        .IncludeTextures();
        }
        
        private static IQueryable<CelestialObject> IncludeTextures(this IQueryable<CelestialObject> @this)
        {
            return @this.Include(s => s.TextureGroup)
                        .Include(s => s.TextureGroup)
                        .Include(s => s.TextureGroup.TextureGroupToTextures.Select(t => t.Texture.TextureType))
                        .Include(s => s.TextureGroup.TextureGroupToTextures.Select(t => t.Texture.TexturePath))
                        .Include(s => s.TextureGroup.TextureGroupToTextures.Select(t => t.Texture.TextureQuality));
        }
        
        private static IQueryable<CelestialObject> IncludeType(this IQueryable<CelestialObject> @this)
        {
            return @this.Include(s => s.CelestialObjectType);
        }
        
        private static IQueryable<CelestialObject> IncludeRingSystem(this IQueryable<CelestialObject> @this)
        {
            return @this.Include(s => s.RingSystem)
                        .Include(s => s.RingSystem.TextureGroup)
                        .Include(s => s.RingSystem.TextureGroup.TextureGroupToTextures.Select(t => t.Texture.TextureType))
                        .Include(s => s.RingSystem.TextureGroup.TextureGroupToTextures.Select(t => t.Texture.TexturePath))
                        .Include(s => s.RingSystem.TextureGroup.TextureGroupToTextures.Select(t => t.Texture.TextureQuality));
        }

    }
}
